### PR TITLE
 Handle reconfiguration for multiple entry of a date type (RWA-2297)

### PR DESCRIPTION
**Pull request by dharmendrak**
### JIRA link (if applicable) ###

[RWA-2297](https://tools.hmcts.net/jira/browse/RWA-2297)

### Change description ###

Handle reconfiguration for multiple entry of a date type

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

**Review by rasikasri — DISMISSED** [review]

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/OriginEarliestDateTypeConfiguratorTest.java [hunk 1]
```diff
@@ -397,4 +452,406 @@ public void shouldCalculateDateWhenOriginDateIsBSTDueDateNonBSTForward() {
...
+            .canReconfigure(CamundaValue.booleanValue(true))
+            .build();
+
+        ConfigurationDmnEvaluationResponse nextHearingDate = ConfigurationDmnEvaluationResponse.builder()
+            .name(CamundaValue.stringValue("nextHearingDate"))
+            .value(CamundaValue.stringValue(nextHearingDateValue + "T21:00"))
+            .canReconfigure(CamundaValue.booleanValue(true))
+            .build();
+        ConfigurationDmnEvaluationResponse dueDate = ConfigurationDmnEvaluationResponse.builder()
+            .name(CamundaValue.stringValue("nextHearingDate"))
```
Comment (PrasadKumarP) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/OriginEarliestDateTypeConfiguratorTest.java:531]: it should be "duedatevalue"

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/NextHearingDateTypeConfiguratorTest.java [hunk 2]
```diff
@@ -1066,4 +1066,75 @@ public void shouldNotDefaultWhenOriginRefAttributesPresentButAreEmpty(boolean co
             .isEmpty();
     }
 
+    /// multiple
+
+
+    @Test
+    public void shouldNotRecalculateDateWhenDueDateIsUnconfigurableButOriginIsConfigurable() {
```
Comment (PrasadKumarP) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/NextHearingDateTypeConfiguratorTest.java:1073]: refactor the method name

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/NextHearingDateTypeConfiguratorTest.java [hunk 3]
```diff
@@ -1066,4 +1066,75 @@ public void shouldNotDefaultWhenOriginRefAttributesPresentButAreEmpty(boolean co
...
+            .build();
+
+        List<ConfigurationDmnEvaluationResponse> configurationDmnEvaluationResponses = dateTypeConfigurator
+            .configureDates(List.of(nextHearingDate, nextHearingDateOrigin), false, true);
+
+        assertThat(configurationDmnEvaluationResponses).isEmpty();
+    }
+
+    @Test
+    public void shouldNotRecalculateDateWhenDueDateIsUnconfigurableButOriginLatestIsConfigurable() {
```
Comment (PrasadKumarP) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/NextHearingDateTypeConfiguratorTest.java:1096]: refactor the method name

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/NextHearingDateTypeConfiguratorTest.java [hunk 4]
```diff
@@ -1066,4 +1066,75 @@ public void shouldNotDefaultWhenOriginRefAttributesPresentButAreEmpty(boolean co
...
+            .build();
+
+        List<ConfigurationDmnEvaluationResponse> configurationDmnEvaluationResponses = dateTypeConfigurator
+            .configureDates(List.of(nextHearingDate, nextHearingDateOrigin), false, true);
+
+        assertThat(configurationDmnEvaluationResponses).isEmpty();
+    }
+
+    @Test
+    public void shouldNotRecalculateDateWhenDueDateIsUnconfigurableButOriginDurationIsConfigurable() {
```
Comment (PrasadKumarP) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/NextHearingDateTypeConfiguratorTest.java:1118]: refactor the method name

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/OriginLatestDateTypeConfiguratorTest.java [hunk 5]
```diff
@@ -398,4 +399,397 @@ public void shouldCalculateDateWhenOriginDateIsBSTDueDateNonBSTForward() {
...
+            .canReconfigure(CamundaValue.booleanValue(true))
+            .build();
+
+        ConfigurationDmnEvaluationResponse nextHearingDate = ConfigurationDmnEvaluationResponse.builder()
+            .name(CamundaValue.stringValue("nextHearingDate"))
+            .value(CamundaValue.stringValue(nextHearingDateValue + "T21:00"))
+            .canReconfigure(CamundaValue.booleanValue(true))
+            .build();
+        ConfigurationDmnEvaluationResponse dueDate = ConfigurationDmnEvaluationResponse.builder()
+            .name(CamundaValue.stringValue("nextHearingDate"))
```
Comment (PrasadKumarP) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/OriginLatestDateTypeConfiguratorTest.java:479]: Is it  "due date" here

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/OriginLatestDateTypeConfiguratorTest.java [hunk 6]
```diff
@@ -398,4 +399,397 @@ public void shouldCalculateDateWhenOriginDateIsBSTDueDateNonBSTForward() {
...
+    }
+
+    @Test
+    public void shouldReCalculateDateWhenBothNextHearingDateAndNextHearingDateOriginLatestAreSetForReconfiguration() {
+        String dueDateValue = GIVEN_DATE.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+        String nextHearingDateValue = GIVEN_DATE.plusDays(2).format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+
+        ConfigurationDmnEvaluationResponse dueDate = ConfigurationDmnEvaluationResponse.builder()
+            .name(CamundaValue.stringValue("nextHearingDate"))
+            .value(CamundaValue.stringValue(dueDateValue + "T18:00"))
```
Comment (PrasadKumarP) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/OriginLatestDateTypeConfiguratorTest.java:509]: is "nextHearingDateValue" here

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/OriginLatestDateTypeConfiguratorTest.java [hunk 7]
```diff
@@ -398,4 +399,397 @@ public void shouldCalculateDateWhenOriginDateIsBSTDueDateNonBSTForward() {
...
+
+        ConfigurationDmnEvaluationResponse nextHearingDateOriginLatest = ConfigurationDmnEvaluationResponse.builder()
+            .name(CamundaValue.stringValue("nextHearingDateOriginLatest"))
+            .value(CamundaValue.stringValue("dueDate"))
+            .canReconfigure(CamundaValue.booleanValue(true))
+            .build();
+
+        ConfigurationDmnEvaluationResponse nextHearingDate = ConfigurationDmnEvaluationResponse.builder()
+            .name(CamundaValue.stringValue("dueDate"))
+            .value(CamundaValue.stringValue(nextHearingDateValue + "T21:00"))
```
Comment (PrasadKumarP) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/OriginLatestDateTypeConfiguratorTest.java:521]: is  it "dueDateValue"

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/OriginLatestDateTypeConfiguratorTest.java [hunk 8]
```diff
@@ -398,4 +399,397 @@ public void shouldCalculateDateWhenOriginDateIsBSTDueDateNonBSTForward() {
...
+            .canReconfigure(CamundaValue.booleanValue(false))
+            .build();
+
+        ConfigurationDmnEvaluationResponse nextHearingDate = ConfigurationDmnEvaluationResponse.builder()
+            .name(CamundaValue.stringValue("nextHearingDate"))
+            .value(CamundaValue.stringValue(nextHearingDateValue + "T21:00"))
+            .canReconfigure(CamundaValue.booleanValue(false))
+            .build();
+        ConfigurationDmnEvaluationResponse dueDate = ConfigurationDmnEvaluationResponse.builder()
+            .name(CamundaValue.stringValue("nextHearingDate"))
```
Comment (PrasadKumarP) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/OriginLatestDateTypeConfiguratorTest.java:605]: is it "due date"

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/OriginLatestDateTypeConfiguratorTest.java [hunk 9]
```diff
@@ -398,4 +399,397 @@ public void shouldCalculateDateWhenOriginDateIsBSTDueDateNonBSTForward() {
...
+    }
+
+    @Test
+    public void shouldNotReCalculateDateWhenBothNextHearingDateAndOriginLatestAreNotSetForReconfiguration() {
+        String dueDateValue = GIVEN_DATE.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+        String nextHearingDateValue = GIVEN_DATE.plusDays(2).format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+
+        ConfigurationDmnEvaluationResponse dueDate = ConfigurationDmnEvaluationResponse.builder()
+            .name(CamundaValue.stringValue("nextHearingDate"))
+            .value(CamundaValue.stringValue(dueDateValue + "T18:00"))
```
Comment (PrasadKumarP) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/calendar/OriginLatestDateTypeConfiguratorTest.java:623]: is it "nextHearingDateValue"

**Review by PrasadKumarP — APPROVED** [review]

**Pull request merged**